worlds/soe/__init__.py

```diff
@@ -0,0 +1,250 @@
+        if type(item) is str:
+            item = self.item_id_to_raw[self.item_name_to_id[item]]
+        return SoEItem(item.name, force_progression or item.progression, self.item_name_to_id[item.name], self.player)
+
+    def create_regions(self):
+        # TODO: generate *some* regions from locations' requirements?
+        r = Region('Menu', None, 'Menu', self.player, self.world)
+        r.exits = [Entrance(self.player, 'New Game', r)]
+        self.world.regions += [r]
+
+        r = Region('Ingame', None, 'Ingame', self.player, self.world)
+        r.locations = [SoELocation(self.player, loc.name, self.location_name_to_id[loc.name], r)
+                       for loc in _locations]
+        r.locations.append(SoELocation(self.player, 'Done', None, r))
+        self.world.regions += [r]
+
+        self.world.get_entrance('New Game', self.player).connect(self.world.get_region('Ingame', self.player))
+
+    def create_items(self):
+        # clear precollected items since we don't support them yet
+        if type(self.world.precollected_items) is dict:
+            self.world.precollected_items[self.player] = []
+        # add items to the pool
+        self.world.itempool += [item for item in
+                                map(lambda item: self.create_item(item, self.restrict_item_placement), _items)]
+
+    def set_rules(self):
+        self.world.completion_condition[self.player] = lambda state: state.has('Victory', self.player)
+        # set Done from goal option once we have multiple goals
+        set_rule(self.world.get_location('Done', self.player),
+                 lambda state: state._soe_has(pyevermizer.P_FINAL_BOSS, self.world, self.player))
+        set_rule(self.world.get_entrance('New Game', self.player), lambda state: True)
+        for loc in _locations:
+            location = self.world.get_location(loc.name, self.player)
+            set_rule(location, self.make_rule(loc.requires))
+            # limit location pool by item type
+            if self.restrict_item_placement:
+                add_item_rule(location, self.make_item_type_limit_rule(loc.type))
+
+    def make_rule(self, requires: typing.List[typing.Tuple[int]]) -> typing.Callable[[typing.Any], bool]:
+        def rule(state) -> bool:
+            for count, progress in requires:
+                if not state._soe_has(progress, self.world, self.player, count):
+                    return False
+            return True
+
+        return rule
+
+    def make_item_type_limit_rule(self, item_type: int):
+        return lambda item: item.player != self.player or self.item_id_to_raw[item.code].type == item_type
+
+    def generate_basic(self):
+        # place Victory event
+        self.world.get_location('Done', self.player).place_locked_item(self.create_event('Victory'))
+        # generate stuff for later
+        self.evermizer_seed = self.world.random.randint(0, 2**16-1)  # TODO: make this an option for "full" plando?
+
+    def post_fill(self):
```

Comment:
I don't get why this is needed? I hope it isn't, as for each SoE world iterate over each location is not very efficient for large multiworlds.

Reply:
`self.restrict_item_placement` is always False at the moment. i used that for testing since item_rules only work for progression. i can get rid of the code before merging if you want.

will also do the other two changes before the merge